src/centroiders.cpp

```diff
@@ -25,6 +25,88 @@ std::vector<Star> DummyCentroidAlgorithm::Go(unsigned char *image, int imageWidt
     return result;
 }
 
+// a poorly designed thresholding algorithm
+int BadThreshold(unsigned char *image, int imageWidth, int imageHeight) {
+    //loop through entire array, find sum of magnitudes
+    int totalMag = 0;
+    for (int i = 0; i < imageHeight * imageWidth; i++) {
+        totalMag += image[i];
+    }
+    return (((totalMag/(imageHeight * imageWidth)) + 1) * 15) / 10;
+}
+
+// a more sophisticated thresholding algorithm, not tailored to star images
+int OtsusThreshold(unsigned char *image, int imageWidth, int imageHeight) {
+    // code here, duh
+    float total = imageWidth * imageHeight;
+    //float top = 255;
+    float sumB = 0;
+    float sum1 = 0;
+    float wB = 0;
+    float maximum = 0;
+    int level = 0;
+    // make the histogram (array length 256)
+    int * histogram = new int[256];
```

Comment:
When you use the `new` keyword, it allocates dynamic memory which is not automatically freed at the end of the function. I.e., this is a memory leak. There are two options:
* At the end of the function, run `delete histogram`, which will free the memory.
* Use the stack instead of dynamic memory: `int histogram[256];` instead of `int *histogram = new int[256];`

The latter is usually preferable, especially on embedded.